app/controllers/search_controller.rb

```diff
@@ -1,50 +1,35 @@
 class SearchController < LoggedAreaController
   include QueryBuilder
 
+  before_action :search_for_dataset, only: [:search]
+  before_action :make_available_to_js, only: [:search]
+
   def search
     @query = params['q'] || ''
     @sort = params['sort']
     @organisation = params['publisher']
     @location = params['location']
-    @search = Dataset.search(search_query(params))
-    @num_results = @search.results.total_count
     @datasets = @search.page(page_number)
-
-    gon.publishers = get_publishers
-    gon.locations = get_locations
   end
 
   def tips
-    render(:layout => "layouts/application")
+    render(layout: 'layouts/application')
   end
 
   private
 
-  def get_publishers
-    results = Dataset.search(publishers_aggregation_query)
-
-    results.response.aggregations.organisations.org_titles.buckets.map do |bucket|
-      "#{bucket[:key]}"
-    end
+  def search_for_dataset
+    @search = Dataset.search(search_query(params))
+    @num_results = @search.results.total_count
   end
 
-  def get_locations
-    results = Dataset.search(locations_aggregation_query)
-
-    results.response.aggregations.locations.buckets.map do |bucket|
-      "#{bucket[:key]}"
-    end
+  def make_available_to_js
+    gon.publishers = Datasets.publishers(publishers_aggregation_query)
```

Comment:
Note that this is making an array of strings available to the javascript in order to populate the publisher and locations autocompletes (can't use ruby objects)